behave/features/steps/specialist_provider.py

```diff
@@ -366,5 +370,64 @@ def step_impl(context):
     assert CASE_SPLIT_TEXT in full_case_log, "Could not find split case in operators dashboard"
 
 
+@step(u'I am on the CSV upload page')
+def step_impl(context):
+    csv_page = f"{CLA_FRONTEND_URL}{CLA_FRONTEND_CSV_URL}"
+    context.helperfunc.open(csv_page)
+    page = CLA_FRONTEND_CSV_URL
+    wait_until_page_is_loaded(page, context)
+    header = context.helperfunc.find_by_xpath("//header/h2[text()='CSV Upload']")
+    assert header is not None
+
+
+@step(u'I select \'Choose file\' and upload a csv file')
+def step_impl(context):
+    input_csv = context.helperfunc.find_by_xpath("//input[@name='csvfile']")
+    # test will fail if file can't be found or upload fails.
+    input_csv.send_keys(os.getcwd() + CLA_SPECIALIST_CSV_UPLOAD_PATH)
+
+
+@step(u'I select the month and year for the uploaded csv file')
+def step_impl(context):
+    # use setter method to set up the datetime in the Helperfunc class
+    context.helperfunc.set_date_now(datetime.datetime.now().date())
+    date_now = context.helperfunc.get_date_now()
+    # Alter date to be the first day of month
+    # Minis one month because current month will not be visible as option
+    new_date = date_now.replace(day=1, month=date_now.month - 1).strftime("%Y-%m-%d")
+    select = Select(context.helperfunc.find_by_xpath(f'//form[@name="csvForm"]/label/select'))
+    assert select is not None
+    try:
+        select.select_by_value(f"string:{new_date}")
+    except StaleElementReferenceException:
+        # Note: always check generated screenshot to ensure that select option is visible and functioning
+        assert False, f"Could not find {new_date} in select options"
 
 
+@step(u'I select the \'upload\' button and check for errors')
+def step_impl(context):
+    button = context.helperfunc.find_by_xpath("//button[@type='submit'][text()='Upload']")
+    assert button is not None
+    button.click()
+
+    # List of errors appear in a dynamic HTML element that contains a header
+    error_header = "//h2[text()='Please correct the following errors:']"
+    # error appears at the top of the page. error is in a list attribute.
+    error_notice = "//div[1]/ul/li"
+
+    context.csv_page = context.helperfunc.find_by_xpath("//*[@id='wrapper']")
+    try:
+        if context.csv_page.find_element_by_xpath(error_notice) or context.csv_page.find_element_by_xpath(error_header):
```

Comment:
if you changed this slightly you could have access to the messages, maybe something like?
`error_text_no_file=""
error_text_bad_file=""
error_text_no_file=context.csv_page.find_element_by_xpath(error_notice).text()
error_text_bad_file=context.csv_page.find_element_by_xpath(error_header).text()
assert False, f"Errors occurred: {}.format`
No idea why it won't do line breaks

Reply:
No sure how it would know which error to print though if it's in an `or` still. So I've changed it to an `if` and `else if` this has the same effect as an `or` and allows better error messages. I've used the idea of getting into the element to get the error text, which is used in the error message. This is much better now.